ARROW-15917: [Java][Docs] Document how to use Flight artifacts (#12820)
apache/arrow: docs/source/java/install.rst

Java Compatibility
------------------

Java modules are currently compatible with JDK 8, 9, 10, or 11, but only JDK 11 is tested in CI.

Installing from Maven
---------------------

By default, Maven will download from the central repository: https://repo.maven.apache.org/maven2/org/apache/arrow/

Configure your pom.xml with the Java modules needed, for example:

> arrow-vector, arrow-memory-netty and flight-core.

Reviewer: ```suggestion
arrow-vector, arrow-memory-netty, and flight-core.
```

Reviewer: That said: I would rather we show a separate POM with the changes needed for Flight, including os-maven-plugin.

Author: Added

                <artifactId>arrow-memory-netty</artifactId>
                <version>${arrow.version}</version>
            </dependency>
        </dependencies>
    </project>

> In case to add Arrow Java Flight dependencies don't forget to
> add `os-maven-plugin` extension. This `os-maven-plugin` generate
> useful platform-dependent project properties such as `os.detected.name`
> or `os.detected.arch` needed by Arrow Java Flight dependencies.

Reviewer: ```suggestion
To use the Arrow Flight dependencies, also add the ``os-maven-plugin``
plugin. This plugin generates useful platform-dependent properties
such as ``os.detected.name`` and ``os.detected.arch`` needed to resolve
transitive dependencies of Flight.
```

    These packages are not official releases. Use them at your own risk.

Arrow nightly builds are posted on the mailing list at `builds@arrow.apache.org`_.

> The artifacts are uploaded to GitHub. For example, for 2022/03/19, they can be found at `Github Nightly`_.

Reviewer: I don't think we need to change the nightly instructions for Flight. We just need to show using os-maven-plugin above.

Reviewer: Ah, I see, thanks.

Maven cannot directly use the artifacts from GitHub.
Instead, install them to the local Maven repository:

1. Decide nightly packages repository to use, for example: https://github.com/ursacomputing/crossbow/releases/tag/nightly-2022-03-19-0-github-java-jars
2. Add packages to your pom.xml, for example: flight-core (it depends on: arrow-format, arrow-vector, arrow-memeory-core and arrow-memory-netty).

.. code-block:: xml

    <properties>
        <maven.compiler.source>8</maven.compiler.source>
        <maven.compiler.target>8</maven.compiler.target>
        <arrow.version>8.0.0.dev254</arrow.version>
    </properties>

    <dependencies>
        <dependency>
            <groupId>org.apache.arrow</groupId>
            <artifactId>flight-core</artifactId>
            <version>${arrow.version}</version>
        </dependency>
    </dependencies>

> 3. Download packages needed to a temporary directory: pom and jar files

Reviewer: ```suggestion
3. Download the necessary pom and jar files to a temporary directory:
```

    $ wget https://github.com/ursacomputing/crossbow/releases/download/nightly-2022-03-19-0-github-java-jars/arrow-memory-netty-8.0.0.dev254.jar
    $ wget https://github.com/ursacomputing/crossbow/releases/download/nightly-2022-03-19-0-github-java-jars/arrow-flight-8.0.0.dev254.pom
    $ wget https://github.com/ursacomputing/crossbow/releases/download/nightly-2022-03-19-0-github-java-jars/flight-core-8.0.0.dev254.pom
    $ wget https://github.com/ursacomputing/crossbow/releases/download/nightly-2022-03-19-0-github-java-jars/flight-core-8.0.0.dev254.jar
    $ tree
    .
    ├── arrow-flight-8.0.0.dev254.pom
    ├── arrow-format-8.0.0.dev254.jar
    ├── arrow-format-8.0.0.dev254.pom
    ├── arrow-java-root-8.0.0.dev254.pom
    ├── arrow-memory-8.0.0.dev254.pom
    ├── arrow-memory-core-8.0.0.dev254.jar
    ├── arrow-memory-core-8.0.0.dev254.pom
    ├── arrow-memory-netty-8.0.0.dev254.jar
    ├── arrow-memory-netty-8.0.0.dev254.pom
    ├── arrow-vector-8.0.0.dev254.jar
    ├── arrow-vector-8.0.0.dev254.pom
    ├── flight-core-8.0.0.dev254.jar
    └── flight-core-8.0.0.dev254.pom

> 4. Install the artifacts to the local Maven repository with **mvn install:install-file**:

Reviewer: ```suggestion
4. Install the artifacts to the local Maven repository with ``mvn install:install-file``:
```

    $ mvn install:install-file -Dfile="$(pwd)/arrow-memory-netty-8.0.0.dev254.pom" -DgroupId=org.apache.arrow -DartifactId=arrow-memory-netty -Dversion=8.0.0.dev254 -Dpackaging=pom
    $ mvn install:install-file -Dfile="$(pwd)/arrow-memory-core-8.0.0.dev254.jar" -DgroupId=org.apache.arrow -DartifactId=arrow-memory-core -Dversion=8.0.0.dev254 -Dpackaging=jar
    $ mvn install:install-file -Dfile="$(pwd)/arrow-memory-netty-8.0.0.dev254.jar" -DgroupId=org.apache.arrow -DartifactId=arrow-memory-netty -Dversion=8.0.0.dev254 -Dpackaging=jar
    $ mvn install:install-file -Dfile="$(pwd)/arrow-flight-8.0.0.dev254.pom" -DgroupId=org.apache.arrow -DartifactId=arrow-flight -Dversion=8.0.0.dev254 -Dpackaging=pom
    $ mvn install:install-file -Dfile="$(pwd)/flight-core-8.0.0.dev254.pom" -DgroupId=org.apache.arrow -DartifactId=flight-core -Dversion=8.0.0.dev254 -Dpackaging=pom
    $ mvn install:install-file -Dfile="$(pwd)/flight-core-8.0.0.dev254.jar" -DgroupId=org.apache.arrow -DartifactId=flight-core -Dversion=8.0.0.dev254 -Dpackaging=jar

5. Validate that the packages were installed:

.. code-block:: shell

>     $ tree /Users/arrow/.m2/repository/org/apache/arrow

Reviewer: ```suggestion
    $ tree ~/.m2/repository/org/apache/arrow
```

    .
    ├── arrow-flight
    │   ├── 8.0.0.dev254
    │   │   └── arrow-flight-8.0.0.dev254.pom
    ├── arrow-format
    │   ├── 8.0.0.dev254
    │   │   ├── arrow-format-8.0.0.dev254.jar
    │   │   └── arrow-format-8.0.0.dev254.pom
    ├── arrow-java-root
    │   ├── 8.0.0.dev254
    │   │   └── arrow-java-root-8.0.0.dev254.pom
    ├── arrow-memory
    │   ├── 8.0.0.dev254
    │   │   └── arrow-memory-8.0.0.dev254.pom
    ├── arrow-memory-core
    │   ├── 8.0.0.dev254
    │   │   ├── arrow-memory-core-8.0.0.dev254.jar
    │   │   └── arrow-memory-core-8.0.0.dev254.pom
    ├── arrow-memory-netty
    │   ├── 8.0.0.dev254
    │   │   ├── arrow-memory-netty-8.0.0.dev254.jar
    │   │   └── arrow-memory-netty-8.0.0.dev254.pom
    ├── arrow-vector
    │   ├── 8.0.0.dev254
    │   │   ├── _remote.repositories
    │   │   ├── arrow-vector-8.0.0.dev254.jar
    │   │   └── arrow-vector-8.0.0.dev254.pom
    └── flight-core
        ├── 8.0.0.dev254
        │   ├── flight-core-8.0.0.dev254.jar
        │   └── flight-core-8.0.0.dev254.pom

> 6. Compile your project like usual with **mvn clean install**.

Reviewer: ```suggestion
6. Compile your project like usual with ``mvn clean install``.
```
